README.md
This is a fork of [Sagar Vinodababu's](https://github.com/sgrvinod) **[PyTorch Tutorial to Image Captioning](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning)**.

Please check out his [original work](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning) and appreciate the effort put into this. He has provided a great tutorial and explantion of his implementation of the [_Show, Attend, and Tell_](https://arxiv.org/abs/1502.03044) paper.

# Contents

[***Why another fork***](#why-another-fork)

[***Workflow***](#workflow)

[***Summary***](#summary)

[***Changelog***](#changelog)

# Why another fork

This fork was created to preserve this code, while addressing compatibility into future versions of PyTorch and other dependent libraries. Moreover, it is this author's intent to commit many quality of life features as it relates to managing network settings and datasets, while allowing for easy implementation using custom datasets. The original [README.md](docs/README_ORIGINAL.md) has been preserved for information; however, the commands given in the original should not be assumed to work. 

# Workflow

The main steps for custom images to a caption generator are as follows: 
1. [Setup environment](#setup-environment)
2. [Collect images & manually create captions](#collect-images-create-captions)
3. [Create a dataset JSON that mirrors the COCO Captions dataset](#create-dataset-json)
4. [Setup dataset for compatibility with original repository](#setup-dataset)
5. [Train custom network](#training)
6. Evaluate trained network (optional)
7. [Generate caption predictions using untrained images](#generate-captions)

## Setup Environment

While this isn't meant to be a guide for setting up a working python environment, I will list the versions of the libraries that are confirmed to work on this fork below. This was created on a Win 11 machine with the following libraries/dependencies:
* PyTorch 1.11.0 + cu113
* Argparse
* (todo - populate this)

## Collect Images Create Captions

This fork expects a dataset in the same format as [MS COCO 2014 caption validation set](https://github.com/tylin/coco-caption). That said, if you're using a pre-made dataset, you may skip to the next section. When using a custom dataset, this fork expects a folder of images, each with a .txt file of the same name in the same folder with the caption. This script currently only supports one caption per image, but modifying it shouldn't be too hard if you need that to suit your dataset. 

## Create Dataset JSON

A script has been provided, [1_dataset_to_json.py](1_dataset_to_json.py) which will create a JSON file compatible with the training for this and the original repository, which one may also use for other repositories expecting the same format. The usage of this script is as follows:

To run the script, use the below command replacing the <ExperimentName> and <AbsoluteImagePath> with values of your choosing. 
```python
python 1_dataset_to_json.py <ExperimentName> <AbsoluteImagePath>
```
You may use a --BYOdataset flag if you're training off a dataset that is already compatible (e.g. MS Coco Captions). In this instance, you will still need to specify the absolute path of your images, as well as the <ExperimentName>, as the script makes a folder structure and settings file for your experiment. 

This script, when first run, will create a directory structure within the root of the repository formatted as below:
```
> root
  > networks
    > <ExperimentName>
      > dataset
      > img_out
      > txt_out
      > checkpoints
```

Two files will be placed in the <ExperimentName> folder:
* <ExperimentName>.json - The JSON file for your dataset that is formatted similarly to MSCOCO Captions
* <ExperimentName>_settings.json - The settings for your network

If you're bringing your own dataset, you do not need to copy any images, only provide the absolute path to them. You should however, make a copy of the .json file and put it in the root of <ExperimentName> while renaming the file to <ExperimentName>.json (i.e. it should have the same name as the directory it is in)

If you're using your own data, you can simply provide a name and the absolute image path (e.g. c:\images). You may edit the <ExperimentName>_settings.json file to suit your network. This will create a JSON file with a 80,10,10 -> train,test,validation split. This number can be changed by editing the [1_dataset_to_json.py](1_dataset_to_json.py) file

## Setup Dataset

Next, run the 2_create_input_files.py file as the following: 
```
python 2_create_input_files.py <ExperimentName>
```

This file is from the author's orignal work, modified to use the values in the <ExperimentName>_settings.json file. It will make a copy of ALL your images inside the <ExperimentName>/dataset directory, along with the captions. Use with caution, as you may fill your hard drive if working with extremely large datasets since it makes a copy of everything. The h5py file allows python to use the data directly from the hard drive as if it's reading it from memory. This was likely done to reduce processing time and memory footprint, especially if you're using the GPU. The original author also resizes the images to 256x256 in this procedure, so the images being fed to your network may be smaller than their originals. This is okay, since you're only generating captions. 

After a successful run, you should see "Reading TRAIN images and captions, storing to file..." for the train, val, and test images. 

## Training

Training can be run using: 
```
python 3_train.py <ExperimentName>
```
The --resume flag will resume training from the best checkpoint, as determined by the BLEU-4 score. The --fineTune flag will unlock some of the layers of the encoder for continued training. The author of the original work suggests running training until no improvement is seen, then re-run training with fine tuning. 

This will run until "epochs" are reached (in the settings file) or if there is no improvement in the BLEU-4 score (also in the settings). The original author describes this in depth. This implementation has not changed that. 

## Evaluation

--This is not implemented yet--

## Generate Captions

For inference, run:
```
python 4_caption.py datasetName
```
The images used for inference can be set in your <ExperimentName>_settings.json file that was created in step 1 under testing>>path as below:
```
    "testing": {
        "path": "d:/temp",
```
You may use the --max_images argument followed by an integer to only make captions for X amount of images. You may use the --txt_only switch to create text captions with no images, otherwise the script will create both text files and images. 

The images will be created in <ExperimentName>/img_out/ and the text files with predicted captions will be in the <ExperimentName>/txt_out.  The text files will be named with the same name as the images they were derived from. 

# Summary

A typical training/inference session would be run with the below commands: 
```
python 1_dataset_to_json.py testNet c:\path_to_images_and_txt_files
(edit ./networks/testNet/testNet_settings.json to suit your data)
python 2_create_input_files.py testNet
python 3_train.py testNet
python 3_train.py testNet --resume --fineTune
python 4_caption.py testNet --max_images 5
```

# Changelog
1. Move settable parameters from individual .py files to JSON file, initialized at 1_dataset_to_json.py
2. Use of PIL.Image in lieu of scipy.misc.imread and scipy.misc.imresize due to depreciation. 


changelog.md
The following is a list of changes from [Sagar Vinodababu's](https://github.com/sgrvinod) "Show, Attend, and Tell | a PyTorch Tutorial to Image Captioning"

1. 


utils.py
import os
import numpy as np
import h5py
import json
import torch
#from scipy.misc import imread, imresize
from PIL import Image
from tqdm import tqdm
from collections import Counter
from random import seed, choice, sample


def create_input_files(dataset, karpathy_json_path, image_folder, captions_per_image, min_word_freq, output_folder,
                       max_len=100):
    """
    Creates input files for training, validation, and test data.

    :param dataset: name of dataset, one of 'coco', 'flickr8k', 'flickr30k'
    :param karpathy_json_path: path of Karpathy JSON file with splits and captions
    :param image_folder: folder with downloaded images
    :param captions_per_image: number of captions to sample per image
    :param min_word_freq: words occuring less frequently than this threshold are binned as <unk>s
    :param output_folder: folder to save files
    :param max_len: don't sample captions longer than this length
    """

    #assert dataset in {'coco', 'flickr8k', 'flickr30k'}

    # Read Karpathy JSON
    with open(karpathy_json_path, 'r') as j:
        data = json.load(j)

    # Read image paths and captions for each image
    train_image_paths = []
    train_image_captions = []
    val_image_paths = []
    val_image_captions = []
    test_image_paths = []
    test_image_captions = []
    word_freq = Counter()

    for img in data['images']:
        captions = []
        for c in img['sentences']:
            # Update word frequency
            word_freq.update(c['tokens'])
            if len(c['tokens']) <= max_len:
                captions.append(c['tokens'])

        if len(captions) == 0:
            continue

        path = os.path.join(image_folder, img['filepath'], img['filename']) if dataset == 'coco' else os.path.join(
            image_folder, img['filename'])

        if img['split'] in {'train', 'restval'}:
            train_image_paths.append(path)
            train_image_captions.append(captions)
        elif img['split'] in {'val'}:
            val_image_paths.append(path)
            val_image_captions.append(captions)
        elif img['split'] in {'test'}:
            test_image_paths.append(path)
            test_image_captions.append(captions)

    # Sanity check
    assert len(train_image_paths) == len(train_image_captions)
    assert len(val_image_paths) == len(val_image_captions)
    assert len(test_image_paths) == len(test_image_captions)

    # Create word map
    words = [w for w in word_freq.keys() if word_freq[w] > min_word_freq]
    word_map = {k: v + 1 for v, k in enumerate(words)}
    word_map['<unk>'] = len(word_map) + 1
    word_map['<start>'] = len(word_map) + 1
    word_map['<end>'] = len(word_map) + 1
    word_map['<pad>'] = 0

    # Create a base/root name for all output files
    #base_filename = dataset + '_' + str(captions_per_image) + '_cap_per_img_' + str(min_word_freq) + '_min_word_freq'
    base_filename = dataset #+ '_' + str(captions_per_image) + '_cap_per_img_' + str(min_word_freq) + '_min_word_freq'

    # Save word map to a JSON
    with open(os.path.join(output_folder, 'WORDMAP_' + base_filename + '.json'), 'w') as j:
        json.dump(word_map, j)

    # Sample captions for each image, save images to HDF5 file, and captions and their lengths to JSON files
    seed(123)
    for impaths, imcaps, split in [(train_image_paths, train_image_captions, 'TRAIN'),
                                   (val_image_paths, val_image_captions, 'VAL'),
                                   (test_image_paths, test_image_captions, 'TEST')]:

        with h5py.File(os.path.join(output_folder, split + '_IMAGES_' + base_filename + '.hdf5'), 'a') as h:
            # Make a note of the number of captions we are sampling per image
            h.attrs['captions_per_image'] = captions_per_image

            # Create dataset inside HDF5 file to store images
            images = h.create_dataset('images', (len(impaths), 3, 256, 256), dtype='uint8')

            print("\nReading %s images and captions, storing to file...\n" % split)

            enc_captions = []
            caplens = []

            for i, path in enumerate(tqdm(impaths)):

                # Sample captions
                if len(imcaps[i]) < captions_per_image:
                    captions = imcaps[i] + [choice(imcaps[i]) for _ in range(captions_per_image - len(imcaps[i]))]
                else:
                    captions = sample(imcaps[i], k=captions_per_image)

                # Sanity check
                assert len(captions) == captions_per_image

                # Read images
                img = Image.open(impaths[i]).convert('RGB')
                #if len(img.shape) == 2:
                #    img = img[:, :, np.newaxis]
                #    img = np.concatenate([img, img, img], axis=2)
                img = img.resize((256, 256))
                img = np.array(img)
                img = img.transpose(2, 0, 1)
                assert img.shape == (3, 256, 256)
                assert np.max(img) <= 255

                # Save image to HDF5 file
                images[i] = img

                for j, c in enumerate(captions):
                    # Encode captions
                    enc_c = [word_map['<start>']] + [word_map.get(word, word_map['<unk>']) for word in c] + [
                        word_map['<end>']] + [word_map['<pad>']] * (max_len - len(c))

                    # Find caption lengths
                    c_len = len(c) + 2

                    enc_captions.append(enc_c)
                    caplens.append(c_len)

            # Sanity check
            assert images.shape[0] * captions_per_image == len(enc_captions) == len(caplens)

            # Save encoded captions and their lengths to JSON files
            with open(os.path.join(output_folder, split + '_CAPTIONS_' + base_filename + '.json'), 'w') as j:
                json.dump(enc_captions, j)

            with open(os.path.join(output_folder, split + '_CAPLENS_' + base_filename + '.json'), 'w') as j:
                json.dump(caplens, j)


def init_embedding(embeddings):
    """
    Fills embedding tensor with values from the uniform distribution.

    :param embeddings: embedding tensor
    """
    bias = np.sqrt(3.0 / embeddings.size(1))
    torch.nn.init.uniform_(embeddings, -bias, bias)


def load_embeddings(emb_file, word_map):
    """
    Creates an embedding tensor for the specified word map, for loading into the model.

    :param emb_file: file containing embeddings (stored in GloVe format)
    :param word_map: word map
    :return: embeddings in the same order as the words in the word map, dimension of embeddings
    """

    # Find embedding dimension
    with open(emb_file, 'r') as f:
        emb_dim = len(f.readline().split(' ')) - 1

    vocab = set(word_map.keys())

    # Create tensor to hold embeddings, initialize
    embeddings = torch.FloatTensor(len(vocab), emb_dim)
    init_embedding(embeddings)

    # Read embedding file
    print("\nLoading embeddings...")
    for line in open(emb_file, 'r'):
        line = line.split(' ')

        emb_word = line[0]
        embedding = list(map(lambda t: float(t), filter(lambda n: n and not n.isspace(), line[1:])))

        # Ignore word if not in train_vocab
        if emb_word not in vocab:
            continue

        embeddings[word_map[emb_word]] = torch.FloatTensor(embedding)

    return embeddings, emb_dim


def clip_gradient(optimizer, grad_clip):
    """
    Clips gradients computed during backpropagation to avoid explosion of gradients.

    :param optimizer: optimizer with the gradients to be clipped
    :param grad_clip: clip value
    """
    for group in optimizer.param_groups:
        for param in group['params']:
            if param.grad is not None:
                param.grad.data.clamp_(-grad_clip, grad_clip)


def save_checkpoint(data_name, epoch, epochs_since_improvement, encoder, decoder, encoder_optimizer, decoder_optimizer,
                    bleu4, is_best, ckptPath):
    """
    Saves model checkpoint.

    :param data_name: base name of processed dataset
    :param epoch: epoch number
    :param epochs_since_improvement: number of epochs since last improvement in BLEU-4 score
    :param encoder: encoder model
    :param decoder: decoder model
    :param encoder_optimizer: optimizer to update encoder's weights, if fine-tuning
    :param decoder_optimizer: optimizer to update decoder's weights
    :param bleu4: validation BLEU-4 score for this epoch
    :param is_best: is this checkpoint the best so far?
    """
    state = {'epoch': epoch,
             'epochs_since_improvement': epochs_since_improvement,
             'bleu-4': bleu4,
             'encoder': encoder,
             'decoder': decoder,
             'encoder_optimizer': encoder_optimizer,
             'decoder_optimizer': decoder_optimizer}
    filename = 'checkpoint_' + data_name + '.pth.tar'
    torch.save(state, ckptPath + '/' + filename)
    # If this checkpoint is the best so far, store a copy so it doesn't get overwritten by a worse checkpoint
    if is_best:
        torch.save(state, ckptPath + '/BEST_' + filename)


class AverageMeter(object):
    """
    Keeps track of most recent, average, sum, and count of a metric.
    """

    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count


def adjust_learning_rate(optimizer, shrink_factor):
    """
    Shrinks learning rate by a specified factor.

    :param optimizer: optimizer whose learning rate must be shrunk.
    :param shrink_factor: factor in interval (0, 1) to multiply learning rate with.
    """

    print("\nDECAYING learning rate.")
    for param_group in optimizer.param_groups:
        param_group['lr'] = param_group['lr'] * shrink_factor
    print("The new learning rate is %f\n" % (optimizer.param_groups[0]['lr'],))


def accuracy(scores, targets, k):
    """
    Computes top-k accuracy, from predicted and true labels.

    :param scores: scores from the model
    :param targets: true labels
    :param k: k in top-k accuracy
    :return: top-k accuracy
    """

    batch_size = targets.size(0)
    _, ind = scores.topk(k, 1, True, True)
    correct = ind.eq(targets.view(-1, 1).expand_as(ind))
    correct_total = correct.view(-1).float().sum()  # 0D tensor
    return correct_total.item() * (100.0 / batch_size)


4_caption.py
import os
import torch
import tqdm
import torch.nn.functional as F
import numpy as np
import json
import torchvision.transforms as transforms
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import skimage.transform
import argparse
#from scipy.misc import imread, imresize
from PIL import Image

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


def caption_image_beam_search(encoder, decoder, image_path, word_map, beam_size=3):
    """
    Reads an image and captions it with beam search.

    :param encoder: encoder model
    :param decoder: decoder model
    :param image_path: path to image
    :param word_map: word map
    :param beam_size: number of sequences to consider at each decode-step
    :return: caption, weights for visualization
    """

    k = beam_size
    vocab_size = len(word_map)

    # Read image and process
    #img = imread(image_path)
    img = Image.open(image_path).convert('RGB')
    #if len(img.shape) == 2:
    #    img = img[:, :, np.newaxis]
    #    img = np.concatenate([img, img, img], axis=2)
    img = img.resize((256, 256))
    img = np.array(img)
    img = img.transpose(2, 0, 1)
    img = img / 255.
    img = torch.FloatTensor(img).to(device)
    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     std=[0.229, 0.224, 0.225])
    transform = transforms.Compose([normalize])
    image = transform(img)  # (3, 256, 256)

    # Encode
    image = image.unsqueeze(0)  # (1, 3, 256, 256)
    encoder_out = encoder(image)  # (1, enc_image_size, enc_image_size, encoder_dim)
    enc_image_size = encoder_out.size(1)
    encoder_dim = encoder_out.size(3)

    # Flatten encoding
    encoder_out = encoder_out.view(1, -1, encoder_dim)  # (1, num_pixels, encoder_dim)
    num_pixels = encoder_out.size(1)

    # We'll treat the problem as having a batch size of k
    encoder_out = encoder_out.expand(k, num_pixels, encoder_dim)  # (k, num_pixels, encoder_dim)

    # Tensor to store top k previous words at each step; now they're just <start>
    k_prev_words = torch.LongTensor([[word_map['<start>']]] * k).to(device)  # (k, 1)

    # Tensor to store top k sequences; now they're just <start>
    seqs = k_prev_words  # (k, 1)

    # Tensor to store top k sequences' scores; now they're just 0
    top_k_scores = torch.zeros(k, 1).to(device)  # (k, 1)

    # Tensor to store top k sequences' alphas; now they're just 1s
    seqs_alpha = torch.ones(k, 1, enc_image_size, enc_image_size).to(device)  # (k, 1, enc_image_size, enc_image_size)

    # Lists to store completed sequences, their alphas and scores
    complete_seqs = list()
    complete_seqs_alpha = list()
    complete_seqs_scores = list()

    # Start decoding
    step = 1
    h, c = decoder.init_hidden_state(encoder_out)

    # s is a number less than or equal to k, because sequences are removed from this process once they hit <end>
    while True:

        embeddings = decoder.embedding(k_prev_words).squeeze(1)  # (s, embed_dim)

        awe, alpha = decoder.attention(encoder_out, h)  # (s, encoder_dim), (s, num_pixels)

        alpha = alpha.view(-1, enc_image_size, enc_image_size)  # (s, enc_image_size, enc_image_size)

        gate = decoder.sigmoid(decoder.f_beta(h))  # gating scalar, (s, encoder_dim)
        awe = gate * awe

        h, c = decoder.decode_step(torch.cat([embeddings, awe], dim=1), (h, c))  # (s, decoder_dim)

        scores = decoder.fc(h)  # (s, vocab_size)
        scores = F.log_softmax(scores, dim=1)

        # Add
        scores = top_k_scores.expand_as(scores) + scores  # (s, vocab_size)

        # For the first step, all k points will have the same scores (since same k previous words, h, c)
        if step == 1:
            top_k_scores, top_k_words = scores[0].topk(k, 0, True, True)  # (s)
        else:
            # Unroll and find top scores, and their unrolled indices
            top_k_scores, top_k_words = scores.view(-1).topk(k, 0, True, True)  # (s)

        # Convert unrolled indices to actual indices of scores
        #prev_word_inds = top_k_words // vocab_size  # (s)
        prev_word_inds = torch.div(top_k_words, vocab_size, rounding_mode='floor')
        next_word_inds = top_k_words % vocab_size  # (s)

        # Add new words to sequences, alphas
        seqs = torch.cat([seqs[prev_word_inds], next_word_inds.unsqueeze(1)], dim=1)  # (s, step+1)
        seqs_alpha = torch.cat([seqs_alpha[prev_word_inds], alpha[prev_word_inds].unsqueeze(1)],
                               dim=1)  # (s, step+1, enc_image_size, enc_image_size)

        # Which sequences are incomplete (didn't reach <end>)?
        incomplete_inds = [ind for ind, next_word in enumerate(next_word_inds) if
                           next_word != word_map['<end>']]
        complete_inds = list(set(range(len(next_word_inds))) - set(incomplete_inds))

        # Set aside complete sequences
        if len(complete_inds) > 0:
            complete_seqs.extend(seqs[complete_inds].tolist())
            complete_seqs_alpha.extend(seqs_alpha[complete_inds].tolist())
            complete_seqs_scores.extend(top_k_scores[complete_inds])
        k -= len(complete_inds)  # reduce beam length accordingly

        # Proceed with incomplete sequences
        if k == 0:
            break
        seqs = seqs[incomplete_inds]
        seqs_alpha = seqs_alpha[incomplete_inds]
        h = h[prev_word_inds[incomplete_inds]]
        c = c[prev_word_inds[incomplete_inds]]
        encoder_out = encoder_out[prev_word_inds[incomplete_inds]]
        top_k_scores = top_k_scores[incomplete_inds].unsqueeze(1)
        k_prev_words = next_word_inds[incomplete_inds].unsqueeze(1)

        # Break if things have been going on too long
        if step > 50:
            break
        step += 1

    i = complete_seqs_scores.index(max(complete_seqs_scores))
    seq = complete_seqs[i]
    alphas = complete_seqs_alpha[i]

    return seq, alphas


def visualize_att(image_path, seq, alphas, rev_word_map, img_name, smooth=True):
    """
    Visualizes caption with weights at every word.

    Adapted from paper authors' repo: https://github.com/kelvinxu/arctic-captions/blob/master/alpha_visualization.ipynb

    :param image_path: path to image that has been captioned
    :param seq: caption
    :param alphas: weights
    :param rev_word_map: reverse word mapping, i.e. ix2word
    :param smooth: smooth weights?
    """
    image = Image.open(image_path)
    image = image.resize([14 * 24, 14 * 24], Image.Resampling.LANCZOS)

    words = [rev_word_map[ind] for ind in seq]

    for t in range(len(words)):
        if t > 50:
            break
        plt.subplot(int(np.ceil(len(words) / 5.)), 5, t + 1)

        plt.text(0, 1, '%s' % (words[t]), color='black', backgroundcolor='white', fontsize=12)
        plt.imshow(image)
        current_alpha = alphas[t, :]
        if smooth:
            alpha = skimage.transform.pyramid_expand(current_alpha.numpy(), upscale=24, sigma=8)
        else:
            alpha = skimage.transform.resize(current_alpha.numpy(), [14 * 24, 14 * 24])
        if t == 0:
            plt.imshow(alpha, alpha=0)
        else:
            plt.imshow(alpha, alpha=0.8)
        plt.set_cmap(cm.Greys_r)
        plt.axis('off')
    #plt.show()
    plt.savefig(f'./networks/{args.datasetName}/img_out/{img_name}')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Show, Attend, and Tell - Tutorial - Generate Caption')

    parser.add_argument("datasetName", help="Enter the name of the dataset/experiment")
    #parser.add_argument('--img', '-i', help='path to image')
    #parser.add_argument('--model', '-m', help='path to model')
    #parser.add_argument('--word_map', '-wm', help='path to word map JSON')
    #parser.add_argument('--beam_size', '-b', default=5, type=int, help='beam size for beam search')
    parser.add_argument('--max_images', help='maximum number of images to process from path', type=int)
    parser.add_argument('--txt_only', action='store_true', help='Create txt files with no images')
    parser.add_argument('--dont_smooth', dest='smooth', action='store_false', help='do not smooth alpha overlay')

    args = parser.parse_args()
    
    with open(f'./networks/{args.datasetName}/{args.datasetName}_settings.json', 'r') as jsonFile:
        setDat = json.load(jsonFile)
    
    args.model = f'./networks/{args.datasetName}/checkpoints/BEST_checkpoint_{args.datasetName}.pth.tar'
    args.word_map = f'./networks/{args.datasetName}/dataset/WORDMAP_{args.datasetName}.json'
    args.beam_size = setDat['testing']['beam_size']

    # Load model
    checkpoint = torch.load(args.model, map_location=str(device))
    decoder = checkpoint['decoder']
    decoder = decoder.to(device)
    decoder.eval()
    encoder = checkpoint['encoder']
    encoder = encoder.to(device)
    encoder.eval()

    # Load word map (word2ix)
    with open(args.word_map, 'r') as j:
        word_map = json.load(j)
    rev_word_map = {v: k for k, v in word_map.items()}  # ix2word
    dictionary = [k for k, v in word_map.items()]
    #print(rev_word_map)
    
    #create list of images to process up to max_images
    fileList = os.listdir(setDat['testing']['path'])
    fileList = sorted(fileList, reverse=True)
    
    imgFiles = []
    for f in fileList:
        if args.max_images != None:
            if len(imgFiles) >= args.max_images:
                break
        if f[-3:].lower() in {'jpg', 'png'}:
            #if not os.path.exists(f'{setDat["testing"]["path"]}/{f[0:-3]}txt'):
            imgFiles.append(f)
    
    print(f'{len(imgFiles)} images running through network)')
    #run images through network
    for img in tqdm.tqdm(imgFiles):
        imgPath = f"{setDat['testing']['path']}/{img}"
        
        # Encode, decode with attention and beam search
        seq, alphas = caption_image_beam_search(encoder, decoder, imgPath, word_map, args.beam_size)
        alphas = torch.FloatTensor(alphas)
        #print(seq)
        
        #convert to plain text
        phrase = ""
        for i in seq:
            phrase += dictionary[i-1] + ' '
        phrase = phrase[0:-1] #remove last space
        phrase = phrase.replace('<start> ', '')
        phrase = phrase.replace(' <end>', '')
        #print(phrase)
        
        #save txt file
        with open(f'./networks/{args.datasetName}/txt_out/{img[0:-3]}txt', 'w') as f:
            f.write(phrase)
        
        if not args.txt_only:
            # Visualize caption and attention of best sequence
            visualize_att(imgPath, seq, alphas, rev_word_map, img, args.smooth)


2_create_input_files.py
import os
import json
import argparse
from utils import create_input_files

parser = argparse.ArgumentParser()
parser.add_argument("datasetName", help="Enter the name of the dataset/experiment")
args = parser.parse_args()

with open(f'./networks/{args.datasetName}/{args.datasetName}_settings.json', 'r') as jsonFile:
    setDat = json.load(jsonFile)    

if __name__ == '__main__':
    # Create input files (along with word map)
    create_input_files(dataset=args.datasetName,
                       karpathy_json_path=setDat['dataset']['JSON'],
                       image_folder=setDat['dataset']['PATH'],
                       captions_per_image=setDat['dataset']['CAPTIONS_PER_IMAGE'],
                       min_word_freq=setDat['dataset']['MIN_WORD_FREQUENCY'],
                       output_folder=f'./networks/{args.datasetName}/dataset',
                       max_len=setDat['dataset']['MAX_CAPTION_LENGTH'])


1_dataset_to_json.py
import os
import json
import random
import argparse

def checkMakeDir(d):    
    if not os.path.exists(d):
        os.mkdir(d)
        print(f'{d} created.')

#parse arguments
parser = argparse.ArgumentParser()
parser.add_argument("datasetName", help="Enter the name of the dataset/experiment")
parser.add_argument("absoluteIMGpath", help="Enter the absolute image path")
parser.add_argument("--BYOdataset", action="store_true", help="Use this switch if your training off a different dataset")
args = parser.parse_args()

#process arguments/globals
args.absoluteIMGpath = args.absoluteIMGpath.replace('\\', '/') #for win users
imgPath = args.absoluteIMGpath
splits = [80, 10, 10] #train, test, val

#create directory structure
dirs = ['./networks', f'./networks/{args.datasetName}', f'./networks/{args.datasetName}/dataset', 
        f'./networks/{args.datasetName}/img_out', f'./networks/{args.datasetName}/checkpoints',
        f'./networks/{args.datasetName}/txt_out']
for d in dirs:
    checkMakeDir(d)
   
jsonFname = f'./networks/{args.datasetName}/{args.datasetName}.json'
settingsjson = f'./networks/{args.datasetName}/{args.datasetName}_settings.json'

if not args.BYOdataset:
    '''CREATE DATASET JSON'''

    imgNames = []
    captions = []
    allFiles = os.listdir(imgPath)
    random.shuffle(allFiles)

    for f in allFiles:
        if f[-3:].lower() in {'jpg', 'png'}:
            if os.path.exists(f'{imgPath}/{f[0:-3]}txt'):
                imgNames.append(f)
                with open(f'{imgPath}/{f[0:-3]}txt') as txtFile:
                    captions.append(txtFile.read())   
    del allFiles

    assert len(imgNames) == len(captions)

    splitIndex = [int((splits[0] / 100) * len(imgNames)), int((splits[1] / 100) * len(imgNames))]
    #splitIndex.append(len(imgNames) - (splitIndex[0] + splitIndex[1]))

    for i in range(len(imgNames)):
        splits.append('train')
        if i >= splitIndex[0]:
            splits[-1] = 'test'
            if i >= splitIndex[0] + splitIndex[1]:
                splits[-1] = 'val'

    images = []

    for i in range(len(imgNames)):
        #make tokens for text
        tokens = captions[i].split(' ')
        #put in sentences List
        sentence = {"tokens":tokens, "raw":captions[i], "imgid":i}
        sentences = [sentence]
        #create image dictionary and put in images list
        images.append({"filepath": '', "filename": imgNames[i], "imgid": i, "split": splits[i], 
                       "sentences": sentences, })

    #create dictionary with image list
    jsonData = {"images": images}
    del images

    #write json    
    with open(jsonFname, 'w') as jsonFile:
        json.dump(jsonData, jsonFile)
    
    print(f'{len(imgNames)} image/txt pairs found.')
    del jsonData
    del imgNames
    del captions

'''CREATE JSON SETTINGS FILE'''

dataset = {
    "JSON": jsonFname,
    "PATH": args.absoluteIMGpath,
    "CAPTIONS_PER_IMAGE": 1,
    "MIN_WORD_FREQUENCY": 0,
    "MAX_CAPTION_LENGTH": 50
}
network = {
    "Network_NAME": args.datasetName,
    "start_epoch": 0,
    "epochs": 120,
    "batch_size": 32,
    "workers": 0,           # for data-loading; right now, only 1 works with h5py
    "encoder_lr": 0.0001,   # learning rate for encoder if fine-tuning
    "decoder_lr": 0.0004,   # learning rate for decoder
    "grad_clip": 5.,        # clip gradients at an absolute value of
    "alpha_c": 1.,          # regularization parameter for 'doubly stochastic attention', as in the paper
    "print_freq": 100,      # print training/validation stats every __ batches
    "no_improvement_quit": 10 #Quit after X amount of epochs with no BLEU improvement
}
testing = {
    "path": args.absoluteIMGpath,
    "beam_size": 5,
}

jsonDATA = {"dataset": dataset, "network":network, "testing":testing}

with open(settingsjson, "w") as jsonFile: 
    json.dump(jsonDATA, jsonFile, indent=4)